src/methods/permutations/permutations.mjs

```diff
@@ -1,26 +1,55 @@
 import { iterableCurry } from '../../internal/iterable';
 import { permutationsSize } from '../../internal/math';
-import { map } from '../$map/map';
-import { range } from '../range/range';
-import { product } from '../product/product';
-
-export function permutations(iterable, r) {
-  const arr = Array.from(iterable);
-  const len = arr.length;
-  r = r === undefined ? len : r;
+
+function swap(arr, aIdx, bIdx) {
+  const temp = arr[aIdx];
+  arr[aIdx] = arr[bIdx];
+  arr[bIdx] = temp;
+}
+
+function shiftToEnd(arr, idx) {
+  const toShift = arr[idx]; // 1, 2, 3
+  const end = arr.length - 1;
+  for (let i = idx; i < end; i++) {
+    arr[i] = arr[i + 1];
+  }
+  arr[arr.length - 1] = toShift;
+}
+
+export function permutations(iterable, k) {
+  const arr = [...iterable];
+
+  k = k === undefined ? arr.length : k;
+
+  if (k < 0) {
+    throw new TypeError('The k argument to permutations cannot be < 0');
+  }
+
   return {
     *[Symbol.iterator]() {
-      if (r > len) return;
-      const toIndex = i => arr[i];
-      for (const indices of product(...map(range(0, r), () => range(0, len)))) {
-        const set = new Set(indices);
-        if (set.size === r) {
-          yield Array.from(map(set, toIndex));
+      const ist = [0]; // i stack
+      let start = 0; // start === ist.length - 1
+
+      while (ist.length > 1 || ist[0] < arr.length) {
+        swap(arr, start, start + ist[start]);
+        if (ist.length === k) {
+          yield arr.slice(0, k);
+          ++ist[start];
+        } else {
+          ist.push(0);
+          start++;
+        }
+
+        while (ist[start] === arr.length - start) {
+          shiftToEnd(arr, start);
+          ist.pop();
+          start--;
+          ist[start]++;
         }
       }
     },
-    getSize() {
-      return permutationsSize(len, r);
+    get size() {
+      return permutationsSize(arr.length, k);
```

Comment:
Function (`iterable.getSize()`) was because it contains computation and thus must be explicit in use: User must call the function, and therefore, when using it twice, they opt to call the function once and assign its result to a variable.

Property access (`iterable.size`) does not discourage user from calling it twice as they assume data access contains little to no computation, resulting in worse performance.

Reply:
I'd rather do one of two things: do the computation up front even before the user attempts to access `size`, or just cache size for them after the first call to the getter. The second approach promises to be fastest.

Reply:
> do the computation up front even before the user attempts to access `size`, or just cache size for them after the first call to the getter

Implementation of `get size` does not appear to do either of these things.

Reply:
I have changed it to cache size.